pkg/certificateprovider/certificate_provider_factory.go

```diff
@@ -0,0 +1,49 @@
+/*
+Copyright The Ratify Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package certificateprovider
+
+import (
+	"fmt"
+)
+
+var certificateProviders = make(map[string]CertificateProvider)
```

Comment:
I feel that if we want a factory, probably the map is mapping from string to a `CertProviderFactory` so that controller can create multiple inline/akv certStores for each type.

But if we restrict to create one certStore for each type, we could have it mapping to `CertificateProvider`, but remove the `CertProviderFactory` interface and make the `Register` take in `name` and `CertificateProvider` directly.

Reply:
thanks @binbin-li , this is interesting. At the moment, we only need one instance of each provider. Each provider can fetch certificates for many Certificate store of the same type. The certificates are cached in certificatesMap, and the same provider can be reused for a different keyvault since the input paramters to GetCertificates will be different.  Lets discuss a bit more offline to make sure we have the right design for all user scenarios.

Reply:
factory removed as discussed. Please take a look when convenient, thanks!